ldap/servers/plugins/sync/sync_refresh.c

```diff
@@ -89,7 +90,7 @@ sync_srch_refresh_pre_search(Slapi_PBlock *pb)
              * when using their changelog mode. As a result, we parse the cookie to handle this
              * shenangians to determine if this is valid.
              */
-            client_cookie = sync_cookie_parse(cookie, &cookie_refresh);
+            client_cookie = sync_cookie_parse(cookie, &cookie_refresh, &allow_openldap_compat);
```

Comment:
I'll check this, I was pretty sure that there is no NULL derefs. Which code paths can this happen in?

Reply:
If sync_cookie_parse is called with a cookie containing 'rid=' and allow_openldap_compat==False, then client_cookie is NULL. Later if this is a refreshPersist, it will dereferenced few lines after '_if (client_cookie->openldap_compat == PR_TRUE) .._'

Reply:
Indeed, running the sync_repl tests shows this deref:
`#0  0x00007fdc9e741494 in sync_srch_refresh_pre_search (pb=0x7fdbe4000b60) at ldap/servers/plugins/sync/sync_refresh.c:109
109	                if (client_cookie->openldap_compat == PR_TRUE) {
[Current thread is 1 (Thread 0x7fdc06ffd700 (LWP 1329659))]
(gdb) where
#0  0x00007fdc9e741494 in sync_srch_refresh_pre_search (pb=0x7fdbe4000b60) at ldap/servers/plugins/sync/sync_refresh.c:109
#1  0x00007fdca21c44ad in plugin_call_func (list=0x24d73d0, operation=403, pb=0x7fdbe4000b60, call_one=0) at ldap/servers/slapd/plugin.c:2002
#2  0x00007fdca21c4316 in plugin_call_list (list=0x24bbef0, operation=403, pb=0x7fdbe4000b60) at ldap/servers/slapd/plugin.c:1945
#3  0x00007fdca21c10a3 in plugin_call_plugins (pb=0x7fdbe4000b60, whichfunction=403) at ldap/servers/slapd/plugin.c:414
#4  0x00007fdca21b2076 in op_shared_search (pb=0x7fdbe4000b60, send_result=1) at ldap/servers/slapd/opshared.c:588
#5  0x00000000004372e8 in do_search (pb=0x7fdbe4000b60) at ldap/servers/slapd/search.c:388
#6  0x0000000000418ea9 in connection_dispatch_operation (conn=0x33c7588, op=0x24968d0, pb=0x7fdbe4000b60) at ldap/servers/slapd/connection.c:659
#7  0x000000000041af0f in connection_threadmain () at ldap/servers/slapd/connection.c:1770
#8  0x00007fdca1ca4b34 in _pt_root () from /lib64/libnspr4.so
#9  0x00007fdca1c374e2 in start_thread () from /lib64/libpthread.so.0
#10 0x00007fdca1b0b6a3 in clone () from /lib64/libc.so.6
(gdb) print client_cookie
$1 = (Sync_Cookie *) 0x0
`

Reply:
Interesting, I'll check this out :)